src/u_websocket.c

```diff
@@ -69,19 +70,26 @@ static int is_websocket_data_available(struct _websocket_manager * websocket_man
 
 static ssize_t read_data_from_socket(struct _websocket_manager * websocket_manager, uint8_t * data, size_t len) {
   ssize_t ret = 0, data_len;
+  int retries = 0;
   
   if (len > 0) {
     do {
-      if (websocket_manager->tls) {
-        data_len = gnutls_record_recv(websocket_manager->gnutls_session, data, (len - ret));
-      } else if (websocket_manager->type == U_WEBSOCKET_SERVER) {
-        data_len = read(websocket_manager->mhd_sock, data, (len - ret));
-      } else {
-        data_len = read(websocket_manager->tcp_sock, data, (len - ret));
+      if (is_websocket_data_available(websocket_manager)) {
+        if (websocket_manager->tls) {
+          data_len = gnutls_record_recv(websocket_manager->gnutls_session, &data[ret], (len - ret));
+        } else if (websocket_manager->type == U_WEBSOCKET_SERVER) {
+          data_len = read(websocket_manager->mhd_sock, &data[ret], (len - ret));
+        } else {
+          data_len = read(websocket_manager->tcp_sock, &data[ret], (len - ret));
+        }
+        if (data_len > 0) {
+          ret += data_len;
+        } else if (data_len < 0) {
+          ret = -1;
+          break;
+        }
       }
-      if (data_len > 0) {
-        ret += data_len;
-      } else if (data_len < 0) {
+      if ((retries++) > U_WS_MAX_READ_RETRIES) {
```

Comment:
I'm not confident with this one, websockets have a huge size limit (2^64 IIRC) so adding this check would break large websocket data.

Have you seen a use case to justify this check?

Reply:
Not really, I added the retries only as a measure to avoid being stuck in the loop forever.
But in practice I never once saw that this check was really required.
I did not take into account how it could break large websocket data, also the amount of retries is a bit random.

Reply:
Then I think we'd be better off without this check then.

Reply:
I agree, I updated this pull request with this check removed